addons/sale_timesheet/models/product.py

```diff
@@ -92,6 +100,14 @@ def _is_delivered_timesheet(self):
         self.ensure_one()
         return self.type == 'service' and self.service_policy == 'delivered_timesheet'
 
+    @api.onchange('service_policy')
+    def _onchange_service_policy(self):
+        if self.service_tracking != 'no' and self.service_policy == 'delivered_timesheet':
+            if self.project_id and not self.project_id.allow_timesheets:
+                self.project_id = False
+            elif self.project_template_id and not self.project_template_id.allow_timesheets:
+                self.project_template_id = False
+
```

Comment:
In order to prevent code duplication and prevent code discrepancies in the future @tlibioulle has found an alternative (Thank you to him 🙂) :
```suggestion
        self.product_tmpl_id._onchange_service_tracking()
```

What do you think?

cc: @mba-odoo

Reply:
As discussed the problem is unfortunately indeed that product_tmpl_id does not have the `current` values of the product.product during the onchange but the one before the edition. In order to still try to avoid code duplicate, what do you think of the proposition I made in the commit aea3b2bfbdca77840f18fb0e1a396012673a2453 I just pushed?

Thanks 🙂

Reply:
@IT-Ideas I have made a small change. I just privatized the get_onchange_service_policy_updates method. Commit [commit](https://github.com/odoo/odoo/pull/70609/commits/30a7aebbb80f91207713481b3a8253d9bccf8eba)
Thanks
cc @mba-odoo

Reply:
Thats what I did too at first as it is usually the case for onchange methods but as we call it through the property product_tmpl_id in ´product.product’ I thought it make more sense to have it public. I wouldn’t have done it if we would have called it directly from within ´product.product’ like it is the case with the usual inheritence in odoo when using inherit and the same class name.

Reply:
As explained [here](https://www.odoo.com/documentation/14.0/developer/reference/orm.html?highlight=inheritance#delegation) is more a composition than an inheritance ´has a´ rather than ´is a´. With that in mind this mean that we will call a private static method from outside the class.

What do you think?

Reply:
Dear @ppr-odoo, you were right as I did not know the private / public aspect in Python is actually inexistant and has been added in odoo in order to avoid exposing def through rpc. I learned something thanks to you 🙂 
Thank you! 

So it is better to have private as you did, you were absolutely right! Sorry for having bothered you!